src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs

```diff
@@ -238,13 +238,17 @@ public void Close()
             return;
         }
 
-        // When no test session is being used we don't share the testhost
+        // When no test session is being used, we don't share the testhost
         // between test discovery and test run. The testhost is closed upon
         // successfully completing the operation it was spawned for.
         //
         // In contrast, the new workflow (using test sessions) means we should keep
-        // the testhost alive until explicitly closed by the test session owner.
-        if (_testSessionInfo == null)
+        // the testhost alive until explicitly closed by the test session owner, but
+        // only if the testhost is part of a test session (i.e. the proxy operation manager
+        // id is valid), since there is the distinct possibility of test session criteria
+        // changing between spawn and discovery/run, causing a new proxy operation manager
+        // to be spawned on demand instead of dequing an incompatible proxy from the pool.
+        if (_testSessionInfo == null || _proxyOperationManager.Id < 0)
```

Comment:
Not sure to fully understand (based on comment). Are you saying that we except the ID of the proxy operation manager to be always negative when we are ready to close?

Reply:
Yep. A negative ID means the proxy operation manager is not part of a test session, and hence can be disposed of here (both no-session discovery/run and on-demand proxy spawn fit in this category). If the ID is non-negative, this means the proxy is managed by a session and it will be closed only when explicitly requested by the session creator.